### PR TITLE
Create a code coverage report (again)

### LINKED ISSUE #3029: Build/CI: Revive the coverage report
We used to have a CI job that generates a test coverage report for Motoko, but lost that ability when switching to the dune build system. It seems that dune now supports that well: https://discuss.ocaml.org/t/bisect-ppx-the-coverage-tool-now-has-excellent-integration-with-dune/6634

So we should add that back. Maybe including a comment on any PR that reports the change of coverage to encourage improvement here.

### PR DESCRIPTION
**Pull request by nomeata**
run `nix-build -A tests.coverage` and enjoy a glance at
`result/index.html`. Fixes #3029.

Will make CI even slower. Oh well. If someone is annoyed, get some funding for a CI machine.

### CODE REVIEW

**Comment by nomeata** [comment]
Shall we merge this?

File: default.nix [hunk 1]
```diff
@@ -279,6 +282,22 @@ rec {
   deser = ocaml_exe "deser" "deser" null;
   candid-tests = ocaml_exe "candid-tests" "candid-tests" null;
 
+  # exectuable built with coverage:
```
Comment (ggreif) [inline default.nix:285]:
```suggestion
  # executable built with coverage:
```

File: test/Makefile [hunk 2]
```diff
@@ -1,6 +1,7 @@
 QUICK_DIRS = fail idl ld mo-idl repl run
 OTHER_DIRS = perf run-deser run-drun trap
 TEST_DIRS = $(QUICK_DIRS) $(OTHER_DIRS)
+SOURCE_PATHS ?= ../src # will be overriden when building coverage report with nix
```
Comment (ggreif) [inline test/Makefile:4]:
```suggestion
SOURCE_PATHS ?= ../src # will be overridden when building coverage report with nix
```

File: test/Makefile [hunk 3]
```diff
@@ -10,6 +11,16 @@ all:
 quick:
 	$(foreach test_dir,$(QUICK_DIRS),$(MAKE_PAR) -C $(test_dir) quick &&) true
 
+coverage:
+	rm -rf _coverage
+	mkdir _coverage
+	export BISECT_FILE=$$PWD/_coverage/bisect; \
+	export SKIP_RUNNING=yes; \
+	export SKIP_VALIDATE=yes; \
```
Comment (ggreif) [inline test/Makefile:19]:
```suggestion
	export BISECT_FILE=$$PWD/_coverage/bisect \
	       SKIP_RUNNING=yes; \
	       SKIP_VALIDATE=yes; \
```

**Review by ggreif — APPROVED** [review]
Cool, then let's get this in!

**Pull request merged**